### ISSUE TITLE
Core mHC Implementation

### ISSUE THREAD

**Issue by mosi-sol** [allenai/OLMoE#43]
## More efficent

Implementing the **mHC-OLMoE** (Manifold-Constrained Hyper-Connections on OLMoE)

This code includes the 
- **Sinkhorn-Knopp** projection to enforce the doubly stochastic constraint and a wrapper class that can integrate with the existing `Olmoe` architecture.

### 1. The Core mHC Implementation

This module replaces the standard residual addition in the OLMoE block.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class SinkhornProjection(nn.Module):
    """Enforces the Doubly Stochastic constraint on a mixing matrix."""
    def __init__(self, iterations=20):
        super().__init__()
        self.iterations = iterations

    def forward(self, W):
        # Ensure non-negativity via sigmoid as per DeepSeek mHC paper
        W = torch.sigmoid(W)
        # Iterative row and column normalization
        for _ in range(self.iterations):
            W = W / (W.sum(dim=-1, keepdim=True) + 1e-12)
            W = W / (W.sum(dim=-2, keepdim=True) + 1e-12)
        return W

class mHC_Block(nn.Module):
    def __init__(self, config, n_streams=4):
        super().__init__()
        self.n = n_streams
        self.d_model = config.hidden_size # e.g., 2048 for OLMoE-1B
        
        # Manifold Parameters (Birkhoff Polytope)
        self.H_res = nn.Parameter(torch.randn(self.n, self.n))
        self.H_pre = nn.Parameter(torch.randn(self.n, 1))
        self.H_post = nn.Parameter(torch.randn(1, self.n))
        
        self.sinkhorn = SinkhornProjection()
        
        # Scaling factor α initialized small for stability
        self.alpha = nn.Parameter(torch.tensor(0.01))

    def forward(self, X_manifold, sublayer_fn):
        """
        X_manifold: [Batch, Seq, n_streams, d_model]
        sublayer_fn: The original OLMoE Attention or MoE layer
        """
        # 1. Project to Manifold
        M_res = self.sinkhorn(self.H_res)
        M_pre = torch.sigmoid(self.H_pre)
        M_post = torch.sigmoid(self.H_post)

        # 2. Residual Mixing: X_mixed = M_res @ X_manifold
        X_mixed = torch.einsum('ij,bsjd->bsid', M_res, X_manifold)

        # 3. Down-project manifold to single stream for OLMoE
        x_input = torch.einsum('ni,bsnd->bsd', M_pre, X_mixed)
        
        # 4. Process through standard OLMoE sublayer (Attention/MoE)
        # This keeps FLOPs the same as original model
        y = sublayer_fn(x_input)

        # 5. Up-project back to manifold
        X_update = torch.einsum('in,bsd->bsid', M_post, y)

        # 6. Manifold addition with learnable scaling
        return X_mixed + self.alpha * X_update

```

---

### 2. The mHC-OLMoE Layer Wrapper

This wrapper shows how to replace the internal logic of an OLMoE Decoder Layer.

```python
class mHCOlmoeDecoderLayer(nn.Module):
    def __init__(self, config, original_layer):
        super().__init__()
        # Use the original Attention and MoE weights
        self.attention = original_layer.self_attn
        self.moe = original_layer.moe
        self.input_layernorm = original_layer.input_layernorm
        self.post_attention_layernorm = original_layer.post_attention_layernorm
        
        # mHC Logic for both sublayers
        self.mhc_attn = mHC_Block(config)
        self.mhc_moe = mHC_Block(config)

    def forward(self, X_manifold, **kwargs):
        # A. Attention Step
        def attn_wrapper(x):
            x = self.input_layernorm(x)
            return self.attention(x, **kwargs)[0]
        
        X_manifold = self.mhc_attn(X_manifold, attn_wrapper)
        
        # B. MoE Step
        def moe_wrapper(x):
            x = self.post_attention_layernorm(x)
            return self.moe(x)
            
        X_manifold = self.mhc_moe(X_manifold, moe_wrapper)
        
        return X_manifold

```

---

### 3. How to Initialize the Model

To start using this with OLMoE, use the following sequence:

1. **Load OLMoE:** Use `AutoModel.from_pretrained("allenai/OLMoE-1B-7B")`.
2. **Initialize Manifold:** Repeat the initial word embeddings  times.
3. **Final Collapse:** At the end of all layers, average the manifold: `x = X_manifold.mean(dim=2)`.

> Some edits required

---

### Summary

* **Situation:** You need a performant MoE that doesn't suffer from training spikes at scale.
* **Complexity:** Standard Hyper-connections can explode signal gain to .
* **Solution:** Use the provided `SinkhornProjection` and `einsum` logic to keep the signal gain bounded at  while expanding the "residual bandwidth" of OLMoE.

/pi/id close-->lotuschain

**Comment by mosi-sol** [comment]
happy 2026
